### PR TITLE
[docs] - Update I/O manager concept page (DOC-129)

### PR DESCRIPTION
**Pull request by erinkcochran87**
## Summary & Motivation

This PR makes a few small changes to the I/O manager concept page. Instead of adding a callout pointing to `deps` at the top of the page, I opted to move a few sections around to bring the **When to use I/O managers** section closer to the top. (ETA: This page definitely needs a larger overhaul structurally, but that's out of scope for this PR.)

Without the context of when to use an I/O manager, it felt a bit odd to add a callout for `deps` at the top. Additionally, as new users will likely see the SDA concept page first (which focuses a lot on `deps` and not so much on I/O) this feels like it could work. Open to suggestions, though!

**Old**:

Screenshot 2024-02-06 at 3 14 55 PM

**New**:

Screenshot 2024-02-06 at 3 16 26 PM

## How I Tested These Changes

👀

### CODE REVIEW

File: docs/content/concepts/io-management/io-managers.mdx [hunk 1]
```diff
@@ -16,7 +16,9 @@ description: I/O Managers determine how to store asset/op outputs and load asset
   </a> guide.
 </Note>
 
-I/O Managers are user-provided objects that store asset and op outputs and load them as inputs to downstream assets and ops. They can be a powerful tool that helps you reduce boilerplate code and easily change where your data is stored.
+I/O managers are user-provided objects that store asset and op outputs and load them as inputs to downstream assets and ops. They can be a powerful tool that reduces boilerplate code and easily changes where your data is stored.
+
+Functions decorated by <PyObject object="asset" decorator/>, <PyObject object="multi_asset" decorator/>, and <PyObject object="op" decorator/> can have parameters and return values that are loaded from and written to persistent storage. <PyObject module="dagster" object="IOManager" pluralize /> let the user control how this data is stored and how it's loaded in downstream ops and assets. For `@asset` and `@multi_asset`, the I/O manager effectively determines where the physical asset lives.
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:21]:
```suggestion
Functions decorated by <PyObject object="asset" decorator/>, <PyObject object="multi_asset" decorator/>, and <PyObject object="op" decorator/> can return values that are written to persistent storage. Downstream assets and ops can have parameters that load the values from persistent storage. <PyObject module="dagster" object="IOManager" pluralize /> let the user control how this data is stored and how it's loaded in downstream ops and assets. For `@asset` and `@multi_asset`, the I/O manager effectively determines where the physical asset lives.
```

i know this is pulled from later in the doc, but I found this sentence a bit hard to parse on first read through

File: docs/content/concepts/io-management/io-managers.mdx [hunk 2]
```diff
@@ -27,35 +29,28 @@ I/O Managers are user-provided objects that store asset and op outputs and load
...
-
-- You store many of your assets in the same location and follow a consistent set of rules to determine the storage path.
-- You want to swap out I/O manager implementations to change how your assets are stored in local, staging, and production environments.
-- Your asset functions load the upstream dependencies into memory in order to do the function computation.
+If you find yourself writing the same code at the start and end of each asset or op to load and store data, you may want to consider factoring that code into an I/O manager. They can be useful in situations where:
 
-If you find yourself writing the same code at the start and end of each asset or op to load and store data, you may want to consider factoring that code into an I/O manager.
+- Many assets are stored in the same location and follow a consistent set of rules to determine the storage path
+- Assets should be stored differently in local, staging, and production environments
+- Aasset functions load the upstream dependencies into memory to do the function computation
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:51]:
```suggestion
- Asset functions load the upstream dependencies into memory to do the computation
```

I definitely wrote the first version of this bullet point but now i have no idea what I was thinking with "function computation". The basic idea I was trying to get across is that if you need to load the upstream value into memory in order to do the computation that's in the asset function, then maybe I/O managers are a good choice.

File: docs/content/concepts/io-management/io-managers.mdx [hunk 3]
```diff
@@ -108,11 +103,16 @@ Consider the following diagram. In this example, a job has two I/O managers, eac
...
 
-Dagster provides out-of-the-box I/O managers for popular storage systems: AWS S3 (<PyObject module="dagster_aws.s3" object="s3_pickle_io_manager" />), Azure Blob Storage (<PyObject module="dagster_azure.adls2" object="adls2_pickle_io_manager" />), GCS (<PyObject module="dagster_gcp.gcs" object="gcs_pickle_io_manager" />), and Snowflake (<PyObject module="dagster_snowflake" object="build_snowflake_io_manager" />) - or you can write your own: either from scratch or by extending the `UPathIOManager` if you want to store data in an `fsspec`-supported filesystem. For a full list of Dagster-provided I/O managers, refer to the [built-in I/O managers list](#built-in-io-managers-list).
+Dagster provides out-of-the-box I/O managers for popular storage systems, such as Amazon S3 and Snowflake, or you can write your own:
+
+- From scratch, or
+- By extending the `UPathIOManager` if you want to store data in an `fsspec`-supported filesystem
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:109]: not quite relevant to this PR, but could we add a bullet that says
"By extending the `DBIOManager` if you want to store data in a database"

**Review by jamiedemaria — COMMENTED** [review]
🥳

File: docs/content/concepts/io-management/io-managers.mdx (hunk 3 above)
Comment (erinkcochran87) [reply docs/content/concepts/io-management/io-managers.mdx:109]: Sure!

**Review by jamiedemaria — APPROVED** [review]
lgtm!
